server.js

```diff
@@ -20,4 +20,57 @@ app.get("/", function (request, response) {
   response.sendFile(__dirname + "/index.html");
 });
 
+app.get('/messages', function (request, response) {
+  response.json(messages)
+})
+
+app.post('/messages', function (request, response) {
+  let newMessage = {};
+
+  if (request.body.from.trim() && request.body.text.trim()) {
+    newMessage.id = messages.length
+    newMessage.from = request.body.from
+    newMessage.text = request.body.text
+    newMessage.timeSent = new Date()
+    messages.push(newMessage)
+    response.json(messages)
+  } else {
+    response.status(400).json("Please fill all fileds")
+  }
+
+})
+
+app.get('/messages/search', function (request, response) {
+  let searchWord = request.query.text
+  let result = messages.filter(item => item.text.toLowerCase().includes(searchWord.toLowerCase()))
+  if (result.length > 0) {
+    response.json(result)
+  } else {
+    response.status(404).json("Nothing was found")
+  }
+})
+
+app.get('/messages/latest', function (request, response) {
+  let latestMessages = messages.splice(-2)
+  response.json(latestMessages)
+})
+
+app.get('/messages/:id', function (request, response) {
+  let id = parseInt(request.params.id)
+  let result = messages.find(item => item.id === id)
+  if (result !== undefined) {
+    response.json(result)
+  } else {
+    response.json("not found")
+  }
+})
+
+app.delete('/messages/:id', function (request, response) {
```

Comment:
There is a problem with your use of findIndex and splice. Try doing a deletion with an ID number that doesn't exist, for example with an array of 3 messages with IDs 0,1,2, try deleting ID 10. It will delete the last message in the array.
If you add some debug lines, using console.log() you can see what is returned when findIndex does not find the ID, and work out the cause of the problem.

Reply:
```suggestion
app.delete('/messages/:id', function (request, response) {
let id = parseInt(request.params.id),
  result = messages.find(item => item.id === id),
      index = messages.indexOf(result);
  if (index > -1){
  messages.splice(index,1)
  response.json(messages)
    }  else {
  response.status(404).json("not found")
}
})
```

Is this the right way to fix this?

Reply:
Yes, this works well. There does seem to be some disagreement between developers as to whether you should use a 404 or 204 status code if someone tries to delete a resource which doesn't exist, but I think 404 is fine. If your interested in the debate feel free to do some googling. This is also a good resource on HTTP status codes, although it doesn't answer this particular question. https://wiki.developer.mozilla.org/en-US/docs/Web/HTTP/Status